File: 8_Dictionaries/julia/compute_hashes.jl
using SHA

const FILES = ["test.jl"]

function sha256_file_no_whitespace(path::AbstractString)::String
    text = read(path, String)
    normalized = replace(text, r"\s+" => "")
    return bytes2hex(sha256(codeunits(normalized)))
end

function build_hashes()::Vector{Tuple{String,Union{String,Nothing}}}
    out = Tuple{String,Union{String,Nothing}}[]
    for fname in FILES
        if isfile(fname)
            push!(out, (fname, sha256_file_no_whitespace(fname)))
        else
            push!(out, (fname, nothing))
        end
    end
    return out
end

function write_hashes(path::AbstractString="file_hashes.json")
    kvs = build_hashes()
    open(path, "w") do io
        println(io, "{")
        for (i, (k, v)) in enumerate(kvs)
            value_str = v === nothing ? "null" : "\"" * v * "\""
            comma = i < length(kvs) ? "," : ""
            println(io, "  \"" * k * "\": " * value_str * comma)
        end
        println(io, "}")
    end
end

if abspath(PROGRAM_FILE) == @__FILE__
    write_hashes()
    println("Wrote file_hashes.json")
end

File: 8_Dictionaries/javascript/compute_hashes.js
import crypto from "node:crypto";
import fs from "node:fs";

const FILES = ["test.js"];

function sha256FileNoWhitespace(path) {
  const text = fs.readFileSync(path, "utf-8");
  const normalized = text.replace(/\s+/g, "");
  return crypto.createHash("sha256").update(normalized, "utf-8").digest("hex");
}

function buildHashes() {
  const out = {};
  for (const fname of FILES) {
    if (fs.existsSync(fname)) out[fname] = sha256FileNoWhitespace(fname);
    else out[fname] = null;
  }
  return out;
}

function writeHashes(path = "file_hashes.json") {
  fs.writeFileSync(path, JSON.stringify(buildHashes(), null, 2));
}

if (import.meta.url === `file://${process.argv[1].replace(/\\/g, "/")}`) {
  writeHashes();
  console.log("Wrote file_hashes.json");
}
